src/main/java/net/fabricmc/loom/task/MigrateMappingsTask.java

```diff
@@ -61,123 +68,130 @@ public void doTask() throws Throwable {
+		} catch (IOException e) {
+			throw new IllegalArgumentException("Could not find mappings for version " + targetMappingsVersion, e);
 		}
+	}
 
+	private static void migrateMappings(Project project, MinecraftMappedProvider minecraftMappedProvider,
+										Path inputDir, Path outputDir, TinyTree currentMappings, TinyTree targetMappings, LoomGradleExtension extension
+	) throws IOException {
 		project.getLogger().lifecycle(":joining mappings");
-		MappingSet mappingSet = new MappingsJoiner(sourceMappings, targetMappings, "intermediary", "named").read();
+		MappingSet mappingSet = new MappingsJoiner(currentMappings, targetMappings,
+						"intermediary", "named").read();
 
 		project.getLogger().lifecycle(":remapping");
-		Mercury mercury = new Mercury();
+		Mercury mercury = SourceRemapper.createMercuryWithClassPath(project, false);
 
-		for (File file : project.getConfigurations().getByName(Constants.MINECRAFT_DEPENDENCIES).getFiles()) {
-			mercury.getClassPath().add(file.toPath());
-		}
-
-		for (File file : project.getConfigurations().getByName("compileClasspath").getFiles()) {
-			mercury.getClassPath().add(file.toPath());
-		}
+		mercury.getClassPath().add(minecraftMappedProvider.MINECRAFT_MAPPED_JAR.toPath());
+		mercury.getClassPath().add(minecraftMappedProvider.MINECRAFT_INTERMEDIARY_JAR.toPath());
 
 		mercury.getClassPath().add(extension.getMinecraftMappedProvider().MINECRAFT_MAPPED_JAR.toPath());
 		mercury.getClassPath().add(extension.getMinecraftMappedProvider().MINECRAFT_INTERMEDIARY_JAR.toPath());
 
-		mercury.getProcessors().add(MercuryRemapper.create(mappingSet));
-
 		try {
-			mercury.rewrite(inputDir.toPath(), outputDir.toPath());
+			mercury.rewrite(inputDir, outputDir);
 		} catch (Exception e) {
 			project.getLogger().warn("Could not remap fully!", e);
 		}
 
-		project.getLogger().lifecycle(":cleaning file descriptors");
-		System.gc();
+		try {
+			mercury.rewrite(inputDir, outputDir);
+		} catch (Exception e) {
+			project.getLogger().warn("Could not remap fully!", e);
+		}
 	}
 
 	public static class MappingsJoiner extends MappingsReader {
-		private final Mappings sourceMappings, targetMappings;
+		private final TinyTree sourceMappings, targetMappings;
 		private final String fromNamespace, toNamespace;
 
-		public MappingsJoiner(Mappings sourceMappings, Mappings targetMappings, String fromNamespace, String toNamespace) {
+		/**
+		 * Say A is the source mappings and B is the target mappings.
+		 * It does not map from intermediary to named but rather maps from named-A to named-B, by matching intermediary names.
+		 * It goes through all of the intermediary names of A, and for every such intermediary name, call it I,
+		 * matches the named mapping of I in A, with the named mapping of I in B.
+		 * As you might imagine, this requires intermediary mappings to be static across all versions, which they are. :volderthonk:
```

Comment:
Does this account for when an Intermediary name doesn't change but its signature does?

Reply:
I can check, but this was the way it worked before, I just documented it.

Reply:
@liach how is this related to mixin mappings? I don't think it affects that

Reply:
don't think this task updates mixin mappings 😓

Reply:
so what is the problem?

Reply:
@Chocohead My guess is that it won't change the name if the signature changes. But that is the way it was before (it matched using EntryTriple) so no point fixing it in this PR.

Reply:
The comment about them remaining static (ignoring the fact stable is arguably a better description to start with) might be best replaced/removed then given it isn't actually true

Reply:
The intermediary names stay static, the descriptors not necessarily.

Reply:
I was trying to say that that fact is what enables the form of migration we do actually do.

Reply:
What the names do in isolation is not especially useful if the descriptors are also important (which the way it is implemented right now they are). Changing it to something which clarifies any Intermediary which changes descriptor will be missed as if it changed name just for the sake of completeness seems reasonable enough. Especially if it doesn't get fixed any time soon.

Reply:
Sure.